pkg/asyncexec/stream.go

```diff
@@ -0,0 +1,60 @@
+package asyncexec
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+)
+
+// StreamCommand executes a given command in a context directory and streams
+// the output to the according stdin/stdout.
+func StreamCommand(workDir string, command string, args []string) error {
+	cmd := exec.Command(command, args...)
+	if workDir != "" {
+		cmd.Dir = workDir
+	}
+
+	acmd := New(cmd, 256)
+	err := acmd.Run(true)
+	if err != nil {
+		return err
+	}
+
+	// done is used to wait for stream readers to finish before exiting the function.
+	done := make(chan struct{})
+
+	go func() {
+		for b := range acmd.StdoutStream() {
+			_, err := os.Stdout.Write(b)
+			if err != nil {
+				// should never happen, just panic.
+				panic("failed to write to Stdout")
+			}
+		}
+		done <- struct{}{}
+	}()
+	go func() {
+		for b := range acmd.StderrStream() {
+			_, err := os.Stderr.Write(b)
+			if err != nil {
+				// should never happen, just panic.
+				panic("failed to write to stderr")
+			}
+		}
+		done <- struct{}{}
+	}()
+
+	// wait for printers to finish
+	defer func() { _, _ = <-done, <-done }()
```

Comment:
TIL.  Very cool. Did you get this from somewhere? I'm interested to learn more about goroutines and channels

Reply:
>Did you get this from somewhere?

Not really, this is a pattern that i used a lot in some personal projects... mainly learned how to use them the hard way (deadlocks, deadlocks everywhere..)

The same result can be achieved by using [`sync.WaitGroup`](https://github.com/golang/go/blob/master/src/sync/waitgroup.go) , maybe it's worth it to do so?

Reply:
I watched these talks about channels/goroutines/concurrency long time ago - i highly recommend them:
- https://www.youtube.com/watch?v=f6kdp27TYZs
- https://www.youtube.com/watch?v=qmg1CF3gZQ0